### PR TITLE
Add block_pair option to  BreakMeshByBlockGenerator

### LINKED ISSUE #17749: Improve the design of using restricted blocks for BreakMeshByBlockGenerator
## Reason

`BreakMeshByBlockGenerator` currently cannot selectively generate interface between specific block pairs. 

## Design

Generate interfaces using block pairs. 

## Impact

New feature.

### PR DESCRIPTION
**Pull request by jiangwen84**
closes #17749

## Reason

block_restricted does not work for BreakMeshByBlockGenerator

## Design

Fix bug. Add option to create two sides interfaces which will be used in thermal and/or mechanical contact problems. 

## Impact

Code improvement.

### CODE REVIEW

File: framework/include/meshgenerators/BreakMeshByBlockGenerator.h [hunk 1]
```diff
@@ -41,20 +41,29 @@ class BreakMeshByBlockGenerator : public BreakMeshByBlockGeneratorBase
...
+  bool findBlockPairs(subdomain_id_type block_one, subdomain_id_type block_two);
+
   /// the mesh to modify
   std::unique_ptr<MeshBase> & _input;
-  /// the blocks to split the mesh on
-  const std::vector<SubdomainID> _block;
+  // vector of subdomains surrounding which interfaces will be generated
+  const std::vector<SubdomainID> _surrounding_blocks;
+  /// vector of subdomain pairs between which interfaces will be generated.
+  const std::vector<std::vector<SubdomainID>> _block_pairs;
```
Comment (vanwdani) [inline framework/include/meshgenerators/BreakMeshByBlockGenerator.h:52]: If you used a std::set<std::pair> instead of std::vector<std::vector> it would prevent duplicate SubdomainID pairs

Comment (jiangwen84) [reply framework/include/meshgenerators/BreakMeshByBlockGenerator.h:52]: good point. would `std::pair` work as a parameter type?

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 2]
```diff
@@ -29,11 +29,12 @@ BreakMeshByBlockGenerator::validParams()
...
-      "The list of subdomain names where the mesh should be split, "
-      "default all blocks.");
+      "surrounding_blocks",
+      "The list of subdomain names surrounding which interfaces will be generated.");
+  params.addParam<std::vector<std::vector<SubdomainID>>>(
+      "block_pairs", "The list of subdomain pairs between which interfaces will be generated.");
   params.addParam<bool>("add_transition_interface",
-                        true,
+                        false,
                         "If true (default) and block is not empty, a special boundary named "
```
Comment (arovinelli) [inline framework/src/meshgenerators/BreakMeshByBlockGenerator.C:38]: remove (default) in the description

File: framework/include/meshgenerators/BreakMeshByBlockGenerator.h (hunk 1 above)
Comment (arovinelli) [reply framework/include/meshgenerators/BreakMeshByBlockGenerator.h:52]: we should also make _surrounding_blocks a std::set<SubdomainID>

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 3]
```diff
@@ -50,22 +51,44 @@ BreakMeshByBlockGenerator::validParams()
...
-    _block_set(_block.begin(), _block.end()),
-    _block_restricted(parameters.isParamSetByUser("block")),
+    _surrounding_blocks(parameters.isParamSetByUser("surrounding_blocks")
+                            ? getParam<std::vector<SubdomainID>>("surrounding_blocks")
+                            : std::vector<SubdomainID>()),
+    _block_pairs(parameters.isParamSetByUser("block_pairs")
+                     ? getParam<std::vector<std::vector<SubdomainID>>>("block_pairs")
+                     : std::vector<std::vector<SubdomainID>>()),
+    _block_pairs_restricted(parameters.isParamSetByUser("block_pairs")),
+    _surrounding_blocks_restricted(parameters.isParamSetByUser("surrounding_blocks")),
```
Comment (arovinelli) [inline framework/src/meshgenerators/BreakMeshByBlockGenerator.C:61]: shouldn't we first initialize the `_surrounding_blocks_restricted` and `_block_pairs_restricted` and then use them to get the right paramters?

File: framework/include/meshgenerators/BreakMeshByBlockGenerator.h (hunk 1 above)
Comment (vanwdani) [reply framework/include/meshgenerators/BreakMeshByBlockGenerator.h:52]: I'm not sure.  It seems like addParam is always taken in as a string and converted based on the type.  However I can't find any example of it within MOOSE or BISON.

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 4]
```diff
@@ -50,22 +51,44 @@ BreakMeshByBlockGenerator::validParams()
...
+  if (_block_pairs_restricted && _surrounding_blocks_restricted)
+    paramError("'surrounding_blocks' and 'block_pairs' can not be used at the same time.");
+
+  if (_block_pairs_restricted)
+    for (auto block_pair : _block_pairs)
+    {
+      if (block_pair.size() != 2)
+        paramError("Each row of 'block_pairs' must have a size of two (block names).");
+      std::copy(block_pair.begin(), block_pair.end(), std::inserter(_block_set, _block_set.end()));
+    }
```
Comment (arovinelli) [inline framework/src/meshgenerators/BreakMeshByBlockGenerator.C:76]: I have a question here? what happens if I input 
```
block_pairs = `1 2
               2 1`
```
I believe we might need to avoid left-right duplicates.
The code should reorder the input such that the first item of the pairs is the one with the lowest id and the second the block with the highest id, and remove duplicates if presents

**Review by arovinelli — CHANGES_REQUESTED** [review]

File: framework/include/meshgenerators/BreakMeshByBlockGenerator.h (hunk 1 above)
Comment (jiangwen84) [reply framework/include/meshgenerators/BreakMeshByBlockGenerator.h:52]: > I'm not sure. It seems like addParam is always taken in as a string and converted based on the type. However I can't find any example of it within MOOSE or BISON.

I tried to switch to `set`, but I got some compiling errors. 

@dschwen do you know if we can use `set` and/or `pair` as `Parameter` type?

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C (hunk 3 above)
Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:61]: looks like `set` and `pair` are not supported. Let us use `vector` for now.

For our case, the code should work fine even if we have duplicated items.

Comment (arovinelli) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:61]: Ok, just use a vector for the input and store them in std::set<std::pair>, so duplicates will be removed automatically

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C (hunk 4 above)
Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:76]: Actually it does not matter whether we have duplicated items like that. In the code, we loop all the element pairs and only duplicate nodes when it can find a matching a block pair. If it finds one, it will not care whether there is another one has the same (or switch left or right) block pairs. 

I tested on the regression test by adding duplicated items. It still pass tests. I hope that means it is working fine.

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C (hunk 3 above)
Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:61]: That works. Do you want to throw an error or just remove the duplication. Like I explained, duplication does not matter. If you think an error is necessary here, I am happy to make that change.

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C (hunk 4 above)
Comment (arovinelli) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:76]: personally, I would prefer making `std::unorderedset<std::pairs<SubdomainID,SubdomainID>> _block_pairs` and then initialize it in the second part of the constructor directly using the parameter, e.g. inside the {}. Then in `findBlockPairs` you actually search for a pair inside `_block_pairs`.  Basically the same you alredy did for `_block_set`

Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:76]: I do not see how it can avoid the duplication like `1 2` and `2 1`. Change to `set` only avoids duplication of a same pair, correct? Do I miss anything?

Comment (arovinelli) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:76]: No, you are not missing anything. When constructing `_block_pairs `  you should always put the block with the lowest id as the first element of the pair and the element with the highest id as the second element of the pair. This will also be consistent with all other routines that already work in this way.

Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:76]: Ok. now I got it. That makes sense. I will make the change.

**Comment by jiangwen84** [comment]
@arovinelli Could you take another look? Thanks.

**Review by arovinelli — COMMENTED** [review]
Can we add a test case were we have a domain like

 ```
1|1|1|1
- - - -
2|3|3|2
- - - -
2|2|2|2
```

with input 
`block_pairs=1 2;`

and see if what it comes out is what we are expecting?

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 5]
```diff
@@ -123,7 +156,13 @@ BreakMeshByBlockGenerator::generate()
...
 
           Elem * current_elem = mesh->elem_ptr(elem_id);
-          if (blockRestrictedElementSubdomainID(current_elem) != reference_subdomain_id)
+          subdomain_id_type block_id = blockRestrictedElementSubdomainID(current_elem);
+
+          if ((block_id != reference_subdomain_id && !_block_pairs_restricted) ||
+              (block_id != reference_subdomain_id &&
+               reference_subdomain_id != Elem::invalid_subdomain_id &&
+               block_id != Elem::invalid_subdomain_id &&
+               findBlockPairs(reference_subdomain_id, block_id) && _block_pairs_restricted))
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:165]: No. It is a bit hard to explain. But it works. I do not want to make too much change to your original code where it is a bit messy to deal with transition boundary and restricted blocks. 

I will leave it to you to redesign your codes, possibly make it cleaner.

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 6]
```diff
@@ -29,15 +29,18 @@ BreakMeshByBlockGenerator::validParams()
...
+      "block_pairs", "The list of subdomain pairs between which interfaces will be generated.");
   params.addParam<bool>("add_transition_interface",
-                        true,
+                        false,
                         "If true (default) and block is not empty, a special boundary named "
                         "interface_transition is generate between listed blocks and other blocks.");
   params.addParam<bool>(
       "split_transition_interface", false, "Whether to split the transition interface by blocks.");
+  params.addParam<bool>(
+      "add_two_sides_interface", false, "Whether to add two sides interface boundaries.");
```
[the remark this answers is not in the document]
Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:43]: Two interfaces are identical. It will be confusing to use master/slave to define them.

**Comment by jiangwen84** [comment]
> Can we add a test case were we have a domain like
> 
> ```
> 1|1|1|1
> - - - -
> 2|3|3|2
> - - - -
> 2|2|2|2
> ```
> 
> with input
> `block_pairs=1 2;`
> 
> and see if what it comes out is what we are expecting?

The original mesh has 20 nodes. After mesh modifier, it has 24 nodes. The nodes between two `1 and 2` element edges are duplicated. I think that is expected, do you agree?

**Comment by arovinelli** [comment]
> > Can we add a test case were we have a domain like
> > ```
> > 1|1|1|1
> > - - - -
> > 2|3|3|2
> > - - - -
> > 2|2|2|2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > with input
> > `block_pairs=1 2;`
> > and see if what it comes out is what we are expecting?
> 
> The original mesh has 20 nodes. After mesh modifier, it has 24 nodes. The nodes between two `1 and 2` element edges are duplicated. I think that is expected, do you agree?

For this specific case, with quad 4 I would expect 22 as the only nodes that you can freely duplicate without disconnecting domain 3 from 1 are the ones on the free surafecs

**Comment by arovinelli** [comment]
> > > Can we add a test case were we have a domain like
> > > ```
> > > 1|1|1|1
> > > - - - -
> > > 2|3|3|2
> > > - - - -
> > > 2|2|2|2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > with input
> > > `block_pairs=1 2;`
> > > and see if what it comes out is what we are expecting?
> > 
> > 
> > The original mesh has 20 nodes. After mesh modifier, it has 24 nodes. The nodes between two `1 and 2` element edges are duplicated. I think that is expected, do you agree?
> 
> For this specific case, with quad 4 I would expect 22 as the only nodes that you can freely duplicate without disconnecting domain 3 from 1 or 2 from 3 are the ones on the free surfaces

Similarly if you do `block_pairs=1 3;` I would expect 21 nodes as you can only duplicate the node in the middle without compromising mesh integrity

**Comment by jiangwen84** [comment]
> > > Can we add a test case were we have a domain like
> > > ```
> > > 1|1|1|1
> > > - - - -
> > > 2|3|3|2
> > > - - - -
> > > 2|2|2|2
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > with input
> > > `block_pairs=1 2;`
> > > and see if what it comes out is what we are expecting?
> > 
> > 
> > The original mesh has 20 nodes. After mesh modifier, it has 24 nodes. The nodes between two `1 and 2` element edges are duplicated. I think that is expected, do you agree?
> 
> For this specific case, with quad 4 I would expect 22 as the only nodes that you can freely duplicate without disconnecting domain 3 from 1 are the ones on the free surafecs

I would say both cases are reasonable. Maybe your case has make more sense for a fracture problem. Let me take a look at the code again.

**Comment by arovinelli** [comment]
> > > > Can we add a test case were we have a domain like
> > > > ```
> > > > 1|1|1|1
> > > > - - - -
> > > > 2|3|3|2
> > > > - - - -
> > > > 2|2|2|2
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > with input
> > > > `block_pairs=1 2;`
> > > > and see if what it comes out is what we are expecting?
> > > 
> > > 
> > > The original mesh has 20 nodes. After mesh modifier, it has 24 nodes. The nodes between two `1 and 2` element edges are duplicated. I think that is expected, do you agree?
> > 
> > 
> > For this specific case, with quad 4 I would expect 22 as the only nodes that you can freely duplicate without disconnecting domain 3 from 1 are the ones on the free surafecs
> 
> I would say both cases are reasonable. Maybe your case has make more sense for a fracture problem. Let me take a look at the code again.

The idea should be that you can duplicate a node if and only if it is connected only to elements with subdomain id present in a single block pair. 
Still, I'm not sure this would satisfy your needs, but if contact is the case it should, but you could come up with corner cases, like you want to separate 1 from both 2 and 3 then my idea wouldn't work

**Comment by arovinelli** [comment]
Not sure about the current implementation, what happens if you do
``` 
block_pairs=`1 2;
                       1 3`?
so you get the right number of nodes?

**Comment by jiangwen84** [comment]
> Not sure about the current implementation, what happens if you do
> 
> ```
> block_pairs=`1 2;
>                        1 3`?
> so you get the right number of nodes?
> ```

I made some changes to the codes. For previous two cases, it creates 22 and 21 nodes, as expected. 

For this case, now the code creates 27 codes. it makes sense to me. What do you think?

**Comment by arovinelli** [comment]
I think here we are hitting a corner case.
Shouldn't this result in 25 nodes (duplications of all the nodes shared between 1 and 2/3 but only once)?
27 means some of the nodes on the seconf line are duplicate more than once, I suspect domains 2 and 3 are now disconnected or there are nodes not belonging to any element

**Comment by jiangwen84** [comment]
> I think here we are hitting a corner case.
> Shouldn't this result in 25 nodes (duplications of all the nodes shared between 1 and 2/3 but only once)?
> 27 means some of the nodes on the seconf line are duplicate more than once, I suspect domains 2 and 3 are now disconnected or there are nodes not belonging to any element

I would not agree on that.  I think 27 is correct for this problem.  2 and 3 are different blocks,  so it makes sense to create extra nodes between them.

**Comment by jiangwen84** [comment]
> > I think here we are hitting a corner case.
> > Shouldn't this result in 25 nodes (duplications of all the nodes shared between 1 and 2/3 but only once)?
> > 27 means some of the nodes on the seconf line are duplicate more than once, I suspect domains 2 and 3 are now disconnected or there are nodes not belonging to any element
> 
> I would not agree on that. I think 27 is correct for this problem. 2 and 3 are different blocks, so it makes sense to create extra nodes between them.

If you want to have 25 nodes,  you should combine block 2 and 3  first.

**Comment by arovinelli** [comment]
> > > I think here we are hitting a corner case.
> > > Shouldn't this result in 25 nodes (duplications of all the nodes shared between 1 and 2/3 but only once)?
> > > 27 means some of the nodes on the seconf line are duplicate more than once, I suspect domains 2 and 3 are now disconnected or there are nodes not belonging to any element
> > 
> > 
> > I would not agree on that. I think 27 is correct for this problem. 2 and 3 are different blocks, so it makes sense to create extra nodes between them.
> 
> If you want to have 25 nodes, you should combine block 2 and 3 first.

To say the truth I'am a bit puzzled 

What should this generate?
```
1|1|1|1
- - - -
2|3|3|2
- - - -
2|2|2|2

block_pairs=`1 2;
             1 3;`
```
?

In terms of nodes my guess would be this 
```
o-o-o-o-o
|1|1|1|1|
o-o-o-o-o

o-o-o-o-o
|2|3|3|2|
o-o-o-o-o
|2|2|2|2|
o-o-o-o-o

```

instead you are saying you are ok with (notes the disconnection between two 2 and 3 that. you didn't asked for at least according to the input)
```
o--o--o--o--o
| 1| 1| 1| 1|
o--o--o--o--o
o-o o-o-o o-o
|2\ /3|3\ /2|
o--o--o--o--o
| 2| 2| 2| 2|
o--o--o--o--o
```

**Comment by jiangwen84** [comment]
> > > > I think here we are hitting a corner case.
> > > > Shouldn't this result in 25 nodes (duplications of all the nodes shared between 1 and 2/3 but only once)?
> > > > 27 means some of the nodes on the seconf line are duplicate more than once, I suspect domains 2 and 3 are now disconnected or there are nodes not belonging to any element
> > > 
> > > 
> > > I would not agree on that. I think 27 is correct for this problem. 2 and 3 are different blocks, so it makes sense to create extra nodes between them.
> > 
> > 
> > If you want to have 25 nodes, you should combine block 2 and 3 first.
> 
> To say the truth I'am a bit puzzled
> 
> What should this generate?
> 
> ```
> 1|1|1|1
> - - - -
> 2|3|3|2
> - - - -
> 2|2|2|2
> 
> block_pairs=`1 2;
>              1 3;`
> ```
> 
> ?
> 
> In terms of nodes my guess would be this
> 
> ```
> o-o-o-o-o
> |1|1|1|1|
> o-o-o-o-o
> 
> o-o-o-o-o
> |2|3|3|2|
> o-o-o-o-o
> |2|2|2|2|
> o-o-o-o-o
> ```
> 
> instead you are saying you are ok with (notes the disconnection between two 2 and 3 that. you didn't asked for at least according to the input)
> 
> ```
> o--o--o--o--o
> | 1| 1| 1| 1|
> o--o--o--o--o
> o-o o-o-o o-o
> |2\ /3|3\ /2|
> o--o--o--o--o
> | 2| 2| 2| 2|
> o--o--o--o--o
> ```

Like I said, if you would like to break the mesh as shown in case I, you need to have block 1 and merge block 2 and 3 (or a single block). On the other hand, when you have 3 blocks and ask for interfaces between 1 2 and 1 3, you ask for that the nodes between 1 and 2 and 1 and 3 should all be duplicated. The consequence of this is to have nodes between 2 and 3. 

The bottom line is, with current codes, you can create interface either like case 1 or 2 based on your blocks and block_pair choice.

**Comment by jiangwen84** [comment]
@arovinelli and I had some offline discussion yesterday. We came to an agreement on which nodes should be duplicated on block junctions. One example with different cases is shown below. 

3_blocks_example

**Comment by jiangwen84** [comment]
@arovinelli could you take another review?

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 7]
```diff
@@ -114,6 +148,37 @@ BreakMeshByBlockGenerator::generate()
...
+              {
+                should_create_new_node = false;
+                break;
+              }
+              if (!should_create_new_node)
+                break;
+            }
+          }
+        }
+
```
Comment (arovinelli) [inline framework/src/meshgenerators/BreakMeshByBlockGenerator.C:181]: I think this can be greatly simplified
the condition for which we decide to add a node is iff the domains attached to current_node are two, and if the two domains attached to current_node are in the block_pair_list then duplicate the node.
This means we can use an ordered set to capture the nodes and check if those two subdomains form a pair that is in the list.
Something like this should do the trick
```
if (_block_pairs_restricted){
   should_create_new_node = false
   std::set<subdomain_id_type> sets_blocks_for_this_node;
   for (auto elem_id = elems.begin(); elem_id != elems.end(); elem_id++)
      sets_blocks_for_this_node.insert(blockRestrictedElementSubdomainID(mesh->elem_ptr(*elem_id))
   if (sets_blocks_for_this_node == 2){
   {
      auto setIt = setOfStr.begin(); 
      if (findBlockPairs(setIt, std::next(setIt.begin(), 1)));
         should_create_new_node = true
   }
}
```

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C (hunk 5 above)
[the remark this answers is not in the document]
Comment (arovinelli) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:197]: I tried, and that if statement  boils down to:
```
          if ((blockRestrictedElementSubdomainID(current_elem) != reference_subdomain_id) ||
              (_block_pairs_restricted & findBlockPairs(reference_subdomain_id, block_id)))
```

**Review by arovinelli — COMMENTED** [review]
I think you should be able to make the changes i suggested and then we should be  good to go

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C (hunk 7 above)
Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:181]: seems like your logic works better. 

I made the suggested change.

**Comment by jiangwen84** [comment]
@dschwen Could you take a final look at this PR? Thanks.

**Review by arovinelli — DISMISSED** [review]
Looks good to me.
All the final changes I requested were incorporated.

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 8]
```diff
@@ -48,21 +52,44 @@ BreakMeshByBlockGenerator::validParams()
...
-    _interface_transition_name(getParam<BoundaryName>("interface_transition_name"))
+    _interface_transition_name(getParam<BoundaryName>("interface_transition_name")),
+    _add_interface_on_two_sides(getParam<bool>("add_interface_on_two_sides"))
 {
+  if (_block_pairs_restricted && _surrounding_blocks_restricted)
+    paramError("'surrounding_blocks' and 'block_pairs' can not be used at the same time.");
+
+  if (_block_pairs_restricted)
+  {
+    for (auto block_pair : getParam<std::vector<std::vector<SubdomainID>>>("block_pairs"))
```
Comment (dschwen) [inline framework/src/meshgenerators/BreakMeshByBlockGenerator.C:67]: We don't have a `std::vector<std::pair<>>` parameter support? Well, then this will have to do.

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 9]
```diff
@@ -48,21 +51,44 @@ BreakMeshByBlockGenerator::validParams()
...
+    {
+      if (block_pair.size() != 2)
+        paramError("Each row of 'block_pairs' must have a size of two (block names).");
+      std::pair<SubdomainID, SubdomainID> pair =
+          std::make_pair(block_pair[0] < block_pair[1] ? block_pair[0] : block_pair[1],
+                         block_pair[0] < block_pair[1] ? block_pair[1] : block_pair[0]);
```
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:73]: and 

```
std::make_pair(std::min(block_pair[0], block_pair[1]), std::max(block_pair[0], block_pair[1]));
```

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 10]
```diff
@@ -48,21 +52,44 @@ BreakMeshByBlockGenerator::validParams()
...
   if (typeid(_input).name() == typeid(DistributedMesh).name())
     mooseError("BreakMeshByBlockGenerator only works with ReplicatedMesh.");
 
   if (!_add_transition_interface && _split_transition_interface)
     paramError("split_transition_interface",
                "BreakMeshByBlockGenerator cannot split the transition interface because "
                "add_transition_interface is false");
+
+  if (_add_transition_interface && _block_pairs_restricted)
+    paramError("add_transition_interface cannot work with 'block_pairs'.");
```
Comment (dschwen) [inline framework/src/meshgenerators/BreakMeshByBlockGenerator.C:92]: You need to have the parameter name as the first argument here!

**Review by dschwen — CHANGES_REQUESTED** [review]

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C (hunk 8 above)
Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C:67]: I guess we don't have the support yet. :-)

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C (hunk 10 above)
Comment (jiangwen84) [reply framework/src/meshgenerators/BreakMeshByBlockGenerator.C]: Got it.

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 11]
```diff
@@ -50,22 +52,46 @@ BreakMeshByBlockGenerator::validParams()
...
+    _block_pairs_restricted(parameters.isParamSetByUser("block_pairs")),
+    _surrounding_blocks_restricted(parameters.isParamSetByUser("surrounding_blocks")),
     _add_transition_interface(getParam<bool>("add_transition_interface")),
     _split_transition_interface(getParam<bool>("split_transition_interface")),
     _interface_transition_name(getParam<BoundaryName>("interface_transition_name")),
-    _add_two_sides_interface(getParam<bool>("add_two_sides_interface"))
+    _add_interface_on_two_sides(getParam<bool>("add_interface_on_two_sides"))
 {
+  if (_block_pairs_restricted && _surrounding_blocks_restricted)
+    paramError("'surrounding_blocks' and 'block_pairs' can not be used at the same time.");
```
Comment (dschwen) [inline framework/src/meshgenerators/BreakMeshByBlockGenerator.C:63]: here

File: framework/src/meshgenerators/BreakMeshByBlockGenerator.C [hunk 12]
```diff
@@ -50,22 +52,46 @@ BreakMeshByBlockGenerator::validParams()
...
+
+  if (_block_pairs_restricted)
+  {
+    for (auto block_pair : getParam<std::vector<std::vector<SubdomainID>>>("block_pairs"))
+    {
+      if (block_pair.size() != 2)
+        paramError("Each row of 'block_pairs' must have a size of two (block names).");
```
Comment (dschwen) [inline framework/src/meshgenerators/BreakMeshByBlockGenerator.C:70]: and here

**Review by dschwen — CHANGES_REQUESTED** [review]
looks like we both missed two more paramErrors

**Review by dschwen — APPROVED** [review]
